### ISSUE TITLE
The future of repository interfaces for sulu models

### ISSUE THREAD

**Issue by alexander-schranz** [sulu/sulu#5931]
# Problem description

In the last time @nnatter and I did discuss many times how a repository interfaces should look like in future for our domain models (contacts, medias, categories, ...) in Sulu, I think we did find a good solution in current projects. In this issue I want to write down my thoughts about how a RepositoryInterfaces for our Models should look like. It is focused focus to be flexible, easy to extend and consistent between all of them.

# Proposed solution

In the following Example the name `Model` could be replaced with  `Contact`, `Media`, `Category` or any other Entity in Sulu.

## A. 💉 Injecting instead of extending the doctrine entity repository

Extending the doctrine entity repository did make us some problems when doctrine add new features like in the past the `count` method we had the rename our `count` method with a [bc break](https://github.com/sulu/sulu/pull/3694/files). Also our class should be not be connected to the infrastructure code and only make this method available which are part of our interface. So we should inject the doctrine repository instead of extending from it. This can be done in repositories the following way:

```php
class ModelRepository implements ModelRepositoryInterface {
    /** @var EntityRepository<ModelInterface> */
    protected $entityRepository;

    /** @var EntityManagerInterface */
    protected $entityManager;
    
    public function __construct(
        EntityManagerInterface $entityManager
    ) {
        $this->entityRepository = $entityManager->getRepository(ModelInterface::class);
        $this->entityManager = $entityManager;
    }
}
```

This means that the Sulu ModelRepository will not be available over `$entityManager->getRepository` and the only correct way to inject the repository is injecting the ModelRepository over its service id or over the ModelRepositoryInterface. 

## B. 🏭 Factory on the Repository

This is currently done by us this way over the persistence bundle and I think I would keep it this way I don't see a need for a custom ModelFactory service. This is very Sulu specific as we use doctrine metadata as to allow override entities.

```php
class ModelRepositoryInterface {
      public function create(): ModelInterface;
}
```

## C. 📗 Add and Remove over Repository

Currently a lot of entityManager->persist and entityManager->remove are done in services. This should be avoided and our repository should have a `add` and a `remove` method:

```php
class ModelRepositoryInterface {
      public function add(ModelInterface $model): void;
      
      public function remove(ModelInterface $model): void;
}
```

The `flush` should be done over middlewares or called manually by the dev. We should not automatical flush something in handler, managers. The "UserInterface" layer should be the one which tells when it is the right time to flush.

## D. ⚙️ Easy extendable criteria and sorts

Currently the methods at many cases are to specific and not extendable with custom additional filters or custom sort methods. With a more general array parameters this can be avoided and a developer can easily extend the parameter. This is something we need to keep Sulu extendable for projects you may want a more typed and specific methods.

```php
class ModelRepositoryInterface {
      public function findOneBy(array $criteria): ?ModelInterface;
     
      /** @return iterable<ModelInterface> */
      public function findBy(array $criteria, array $sort = []): iterable;

      public function countBy(array $criteria): int;
}
```

## E. 🌊 Add getOneBy and findOneBy methods

There is a common pattern in Java Repositories with  `get` and `find` :

 - `get`: will `throw ModelNotFoundException` when model with given criteria does not exist
 - `find` will `return null` when model with given criteria does not exist

See here also the Java Spring Framework Repositories:

 - https://docs.spring.io/spring-data/jpa/docs/current/api/org/springframework/data/jpa/repository/JpaRepository.html#getOne-ID-
  - https://docs.spring.io/spring-data/commons/docs/current/api/org/springframework/data/repository/query/QueryByExampleExecutor.html?is-external=true#findOne-org.springframework.data.domain.Example-
  
```php
class ModelRepositoryInterface {
      /** @throws ModelNotFoundException */
      public function getOneBy(array $criteria): ModelInterface;
      
      public function findOneBy(array $criteria): ?ModelInterface;
}
```

Why both? Where in most cases you want that a `ModelNotFoundException` is thrown there are some cases where this make the code a little bit hard to read. Example for validation you want to check if a entity with a specific unique code exist in your repository and want to throw a specific already exist exception.

With the `getOneBy` this would look like:

```php
try {
      $model = $this->modelRepository->getOneBy(['code' => $code]);
      
      throw new ModelAlreadyExistException(['code' => $code]);
} catch (ModelNotFoundException $e) {
      // @ignoreException do nothing expected to get no model here
}
```

Adding a nullable return method like the following make this code a lot easier:

```php
$model = $this->modelRepository->findOneBy(['code' => $code]);

if ($model) {
    throw new ModelAlreadyExistException(['code' => $code]);
}
```

## F. ⛓️ Add `$with` parameter to avoid over and underfetching

> See [`UPDATE 2021-04-19`](#update-2021-04-19) for an alternate version for extending.
> Not yet 100% happy with it (2024-11-20)

Currently there is no way in our repository to tell that you want to fetch also a relation. In some cases example the media a lot of unnecessary joins are done in cases where you only need specific data from the model and not the model with all its subentities. By default our repositories should select only the `main` entity and so only return that one and not add additional joins which are not needed. With the additional `$withs` parameter the developer can add that additional sub entities should also be joined and selected.

```php
class ModelRepositoryInterface {
      public function getOneBy(array $criteria, array $with = []): ModelInterface;
      
      public function findOneBy(array $criteria, array $with = []): ?ModelInterface;
     
      /** @return iterable<ModelInterface> */
      public function findBy(array $criteria, array $sort = [], array $with = []): iterable;
}
```

Example calls:

```php
$modelRepository->getOneBy(['id' => 1], ['addresses' => true]);
// or
$modelRepository->findOneBy(['key' => 'my_key'], ['translations' => ['locale' => en]]);
```

## G. ↕️ Sort By Ids in the repository by sort parameter

Currently sorting by ids are done in different services like managers, controllers, content types. Sorting by ids should be the same like sorting for any other column and should just be given as sort parameter. 

Example calls:

```php
// the following will sort by the given ids
$modelRepository->findBy(['ids' => [2, 1, 7]], ['ids' => 'ASC']); // result 2, 1, 7
// the following will do no sorting and just return like the database would return it
$modelRepository->findBy(['ids' => [2, 1, 7]]); // result order unkown
// the following will sort by
$modelRepository->findBy(['ids' => [2, 1, 7]], ['id' => 'ASC']); // result 1, 2, 7
```

So there need to be a difference between sort by `id` and sort by `ids` think could be a little bit hard to read it could be discussed if it should be `'given-ids' => 'ASC'`. Maybe for tag something like `'given-names' => 'ASC'` could exist.

Where at first look its maybe strange to have `ASC` and `DESC` here but in some kind of UI the user could sort forward and backward by ids where this could be needed. 

## H.  📋 Add flat method to the repository

At current state I mostly multiple services to create a controller for a basic entity like the DoctrineFieldDescriptor factory. This should not be the case as the repository should be the source also for providing a flat list.

So the ModelRepository should have a `findFlatBy` method to get a flat list of models:

```php
class ModelRepositoryInterface {
      /** @return array<array<string, mixed>> */
      public function findFlatBy(array $fields = [], array $sortBy = []): array;
}
```

In the implementation the listBuilder and FieldDescriptorFactory should be used.

## I.  🙋 Add existBy method to the repository **UPDATE 2022-03-29** <a href="#update-2022-03-29" id="update-2022-03-29">#</a>:

Add an `existBy` method to the RepositoryInterface to check if a model with the same value already exist:

```php
class ModelRepositoryInterface {
      public function existBy(array $criteria = []): bool;
}
```

This can as example used to check if a user already exist with that email address e.g. `$this->existBy(['email' => $email]);`, the advantages to `findBy` or `findOneBy` is that
there is no need in this cases to hydrate an object as we only need to know if it exist or not to show an error or throw an exception:

```php
if ($this->userRepository->existBy(['email' => $email])) {
    throw new UserAlreadyExistException($email);
}
```

# Example Repository

So with all proposal above a Repository should look like the following:

```php
class ModelRepositoryInterface {
      public function create(): ModelInterface;

      public function add(ModelInterface $model): void;
     
      public function remove(ModelInterface $model): void;
     
      public function getOneBy(array $criteria, array $with = []): ModelInterface;
      
      public function findOneBy(array $criteria, array $with = []): ?ModelInterface;
     
      /** @return iterable<ModelInterface> */
      public function findBy(array $criteria, array $sort = [], array $with = []): iterable;
     
      public function countBy(array $criteria): int;
      
      public function existBy(array $criteria): bool;

      /** @return array<array<string, mixed>> */
      public function findFlatBy(array $fields = [], array $sortBy = []): array;
}
```

```php
class ModelRepository implements ModelRepositoryInterface {
    /** @var EntityRepository<ModelInterface> */
    protected $entityRepository;

    /** @var EntityManagerInterface */
    protected $entityManager;
    
    /**
     * @var string
     */
    protected $className;
    
    public function __construct(
        EntityManagerInterface $entityManager
    ) {
        $this->entityRepository = $entityManager->getRepository(ModelInterface::class);
        $this->className = $this->entityRepository->getClassName();
        $this->entityManager = $entityManager;
    }

      public function create(): ModelInterface
      {
           return new $this->className();
      }

      public function add(ModelInterface $model): void
      {
           $this->entityManager->persist($model);
      }
     
      public function remove(ModelInterface $model): void
      {
           $this->entityManager->remove($model);
      }

      public function findFlatBy(array $fields = [], array $sortBy = []): array
      {
           // do listbuilder stuff and set $fields and $sortBy
      
           return $listBuilder->execute();
      }

      public function getOneBy(array $criteria, array $with = []): ModelInterface;
      {
            $model = $this->findOneBy($criteria, $with);
            
            if (!$model) {
                  return ModelNotFoundException($criteria);
            }
            
            return $model;
      }
      
      public function findOneBy(array $criteria, array $with = []): ?ModelInterface
      {
            $queryBuilder = $this->createQueryBuilder($criteria, [], $with);
            
            return $queryBuilder->getQuery->getSingleResultOrNull();
      }
     
      /** @return iterable<ModelInterface> */
      public function findBy(array $criteria, array $sort = [], array $with = []): iterable
      {
            $queryBuilder = $this->createQueryBuilder($criteria, $with);
            
            $result = $queryBuilder->getQuery->getResult();;

            $sortByIds = $sort['ids'] ?? null;
            if ($sortByIds) {
                  // sort by ids here
            }
            
            // in best cases this should be a generator for performance
            return $result;
      }
     
      public function countBy(array $criteria): int
      {
            // paginations filters should have no effect on countBy
            unset($criteria['limit']);
            unset($criteria['offset']);

            $queryBuilder = $this->createQueryBuilder($criteria, $with);
            $queryBuilder->select('COUNT(model)');
            
            return (int) $queryBuilder->getQuery->getSingleScalarResult();
      }
     
      public function existBy(array $criteria): int
      {
            $queryBuilder = $this->createQueryBuilder($criteria);
            $queryBuilder->select('model.id')
                ->setMaxResults(1);
            
            return count($queryBuilder->getQuery()->getArrayResult()) > 0;
      }
      
      protected function createQueryBuilder(array $criteria, array $sort = [], array $with = [])
      {
            $queryBuilder = $this->enityRepository->createQueryBuilder($criteria, $with);

             // filters
             if (\array_key_exist('id', $data)) {
                    $queryBuilder->andWhere('model.key = (:key)')
                    $queryBuilder->setParameter('id', $data['id']);
            }
            
             if (\array_key_exist('key', $data)) {
                    $queryBuilder->andWhere('model.key = (:key)')
                    $queryBuilder->setParameter('key', $data['key']);
            }
            
            $ids = $criteria['ids'] ?? [];
            if (!empty($ids)) {
                    $queryBuilder->andWhere('model.id IN (:ids)')
                    $queryBuilder->setParameter('ids', $ids);
            }

            $search = $criteria['search'] ?? null;
            if ($search) {
                    $queryBuilder->leftJoin('model.translations', 'translation', JOIN::WITH, 'translation.locale = :locale')
                    $queryBuilder->andWhere('translation.title LIKE (:search)')
                    $queryBuilder->setParameter('search', '%' . $search . '%');
                    $queryBuilder->setParameter('locale', $data['locale']);
            }
            
            $limit = $criteria['limit'] ?? null;
            if ($limit) {
                   $queryBuilder->setMaxResults($limit);
            }
            
            $offset = $criteria['offset'] ?? null;
            if ($offset) {
                   $queryBuilder->setFirstResult($offset);
            }
            
            // sortBy
            foreach ($sort as $sortByField => $sortByOrder) {
                    $queryBuilder->sortBy('model.' . $sortByField, sortByOrder);
            }
            
            // withs
            $withTranslations = $criteria['translations'] ?? null;
            if ($withTranslations) {
                    $queryBuilder->leftJoin('model.translations', 'translation');
                    
                    if ($withTranslations['locale'] ?? null) {
                         $queryBuilder->setParameter('locale', $withTranslations['locale']);
                    }
            }

            return $queryBuilder;
      }
}
```

# Extended Example

> See [`UPDATE 2021-04-12`](#update-2021-04-12) for an alternate version for extending.

As we should also focus what is the right way to extend the repositories filters and sorts. The following way would be my recommended way then to extend the createQueryBuilder to add additional `filters`, `sorty` or `with` joins.

```php
class CustomModelRepository extends ModelRepository {
      protected function createQueryBuilder(array $criteria, array $sort = [], array $with = [])
      {
             $queryBuilder = parent::createQueryBuilder($criteria, $sort, $with);
             
             if (\array_key_exist('customField', $data)) {
                    $queryBuilder->andWhere('model.customField = :customField');
                    $queryBuilder->setParameter('customField' ,$data['customField']);
             }
             
             return $queryBuilder;
      }
}
```

Mostly there are 2 ways of overriding a service in symfony, the first and recommended way would be decorating the service. Decorating is in this case not possible as the createQueryBuilder is not a method called from outside and is also a Infrastructure specific method. So extending is here the only way to make the repository querybuilder method overwriteable. This requires us in the repositories to have always a bc promise for the constructor.

--------------------

**UPDATE 2021-04-12**  <a href="#update-2021-04-12" id="update-2021-04-12">#</a>:

## Alternative usage of Visitor Pattern as extension point

As mention in some comments extend can make it hard for bundle developer extending things. They would need to provide interface and trait which the project need to implement thereself. As an alternative a [`Visitor Pattern`](https://en.wikipedia.org/wiki/Visitor_pattern) could be used for the QueryBuilder extension point.

This would require an additional interface 

```php
interface ModelQueryBuilderVisitorInterface {
     public function visitQueryBuilder(QueryBuilder $queryBuilder, array $criteria, array $sort = [], array $with = []): void;
}
```

Service implementing it would needed to be tagged with something like `sulu_bundle.model_doctrine_query_builder_visitor` that all instance of it can be injected automatically in the Doctrine ModelRepository:

```php
class ModelRepository implements ModelRepositoryInterface {
    // ..
     
    /** ModelQueryBuilderVisitorInterface[] */
    private $modelQueryBuilderVisitors = [];
      
    public function __construct(
        // ...
        $modelQueryBuilderVisitors = [];
    ) {
        // ...
        $this->modelQueryBuilderVisitors = $modelQueryBuilderVisitors;
    }

     protected function createQueryBuilder(array $criteria, array $sort = [], array $with = [])
     {
           $queryBuilder = ...;
      
           foreach ($this->modelQueryBuilderVisitors[] as $queryBuilderVisitor) {
                  $queryBuilderVisitor->visitQueryBuilder($queryBuilder, array $criteria, array $sort = [], array $with = []);
           }
     }
}
```

**UPDATE 2021-04-19** <a href="#update-2021-04-19" id="update-2021-04-19">#</a>:

## Alternative usage of `$context` instead of `$with`

The `$with` come with one disadvantage that the caller need to know which one need to be loaded. In sulu where in the api's the caller is a controller and if somebody extend the entity we want to avoid that they need to extend also the controller, so the `$with` can be a problem here. So the alternative here would be that we reuse the `$names` of the serialization contexts which we use to serialize the objects:

```php
class ModelRepositoryInterface {
      public function getOneBy(array $criteria, array $contexts = []): ModelInterface;
      
      public function findOneBy(array $criteria, array $contexts = []): ?ModelInterface;
     
      /** @return iterable<ModelInterface> */
      public function findBy(array $criteria, array $sort = [], array $contexts = []): iterable;
}
```

Before with `$with`:

```php
$modelRepository->getOneBy(['id' => 1], ['addresses' => true]);
// or
$modelRepository->findOneBy(['key' => 'my_key'], ['translations' => ['locale' => 'en']]);
```

Now with `$context`:

```php
$modelRepository->getOneBy(['id' => 1], ['modelFullAdminApi']);
// or
$modelRepository->findOneBy(['key' => 'my_key', 'locale' => 'en'], ['modelFullAdminApi']);
```

The locale is here correctly a filter param.

Internal the `context` could still work the same like the `$with` mechanism as an example:

```php
protected function getContextConfiguration(): array
{
       return [
               'modelMinimalAdminApi' => [
                     'with-translations' => true,
               ],
               'modelFullAdminApi' => [
                     'with-translations' => true,
                     'with-addresses' => true,
               ],
       ];
}

private function getWith(array $contexts = []): array
{
       $configuredContexts = $this->getContextConfiguration();
       
       $with = [];
       foreach ($contexts as $context) {
             $with[] = \array_merge($with, $configuredContexts[$context]);
       }
       
       return $with;
}
```

So in this case instead of that they need also extend the caller they can just add additional joins to a specific serialization context. And can create new contexts. Not sure if the option with `with` should still exist. There are usecases where `$with` case makes more sense then the `$context` and the other way around. Having both `$with` and `$context` is conflicting. And adding a class for it make the usage of it harder specially if you want to provide twig methods for it. 

The `$context` could also be extendable over the visitor pattern, see for this the alternative on the UPDATE 2021-04-12.

**Comment by gggeek** [comment]
What about adding offset/limit to the `find` methods?

**Comment by lchrusciel** [comment]
Hey Alexander,

thanks for sharing your perspective over repositories extendibility. I don't want to interfere too much with your perspective(however from my experience inheritance starts to be a real pain when we have additional bundles/plugins on top of the framework, but below third-party app), but I would like to suggest putting what you've written into ADR as a part of the project in MD file. Therefore people may find it even in the vendor folder and there is the whole story in the git history. For the last year, we are doing it in Sylius and I can still recommend this approach.

**Comment by alexander-schranz** [comment]
@gggeek I know that some people prefer having them as own arguments in the methods, but from my perspective there is no difference to other filter parameters. As a parameter like limit, page (offset) are filtering the resultset like other filters. That's why the are part of the $criteria, this make it from the usage inside controller in combination with other filters easier as you can directly forward `$request->query->all()` to the repository. Maybe I will rename the `$criteria` to `$filters` so this is don't get to much associated with doctrine criterias. Thx for your feedback.

**Comment by alexander-schranz** [comment]
@lchrusciel Thx for your feedback. I aggree that extending then in third party and plugins could be difficult, but they are I think also difficult on the Model / Entity level. If there are 2 plugins/bundles adding a new field to an entity you run mostly in the same problem. Currently on entity level I only see that then the bundles would need to provide an own interface and maybe a trait, which then can be used in the project. I think that would be the same case for a repository. How are you handling this in sylius with plugins extending entities?

I'm also thinking at current feedback maybe make usage of the [visitor pattern](https://en.wikipedia.org/wiki/Visitor_pattern) for the Repository as extension point. This would look something like this:

This would require a additional interface 

```php
interface ModelQueryBuilderVisitorInterface {
     public function visitQueryBuilder($queryBuilder, array $criteria, array $sort = [], array $with = []): void;
}
```

Service implementing it would needed to be tagged with something like `sulu_bundle.model_doctrine_query_builder_visitor` that all instance of it can be injected automatically in the Doctrine ModelRepository:

```php
class ModelRepository implements ModelRepositoryInterface {
    // ..
     
    /** ModelQueryBuilderVisitorInterface[] */
    private $modelQueryBuilderVisitors = [];
      
    public function __construct(
        // ...
        $modelQueryBuilderVisitors = [];
    ) {
        // ...
        $this->modelQueryBuilderVisitors = $modelQueryBuilderVisitors;
    }

     protected function createQueryBuilder(array $criteria, array $sort = [], array $with = [])
     {
           $queryBuilder = ...;
      
           foreach ($this->modelQueryBuilderVisitors[] as $queryBuilderVisitor) {
                  $queryBuilderVisitor->visitQueryBuilder($queryBuilder, array $criteria, array $sort = [], array $with = []);
           }
     }
}
```

**Comment by lchrusciel** [comment]
In Sylius, entities are extended with traits with default Doctrine mapping - solution good enough for now but can be felt that it may backfire in the long run. The only other possibility is splitting entities into smaller chunks to avoid this issue as much as possible, but it is not always possible.

The repositories, for now, are extended mostly just by extending the previous class. However, it leads maybe not to inheritance hell, but it is not pleasant to work with. Also, it makes problems to the Liskov substitution principle, as some interfaces tend to grow pretty big. Once you would like to change the repository logic, you end up with a lot of "MethodNotImplementedException". In some cases, we've introduced classes focused on only one job, which makes it much easier to substitute just part of the system, and it seems to be a better approach than the previous one. As you mention, repositories and entities may suffer similar issues, but it seems much easier to fix it on repositories. Easy enough that it is worth pursuing it. 

Visitor pattern may also be a feasible option. I didn't consider it yet. I have to thank you for sharing your ideas once again ;)

**Comment by alexander-schranz** [comment]
@lchrusciel thank you for your response. I understand your point and maybe something like the visitor pattern would make it better to work and better extendable for third party libraries.

I also to rethink about the `$with` that it has some disadvantages as the caller of the method need to know which one need to be loaded and maybe a better option would be instead using `$contexts` and reuse serialization name context in the repository, so a connection between serialization and which is needed to be loaded for the serialization context is there. I added an example in the Issue Description on the [`2021-04-19` update](#update-2021-04-19).

**Comment by alexander-schranz** [comment]
Added a new updated version which adds a new method to the Repository [the `existBy` Update 2022-03-29](https://github.com/sulu/sulu/issues/5931#update-2022-03-29) under the new point `I.` in the issue description and its usecases. The example repository is also updated with it.

**Comment by alexander-schranz** [comment]
Found another interesting Repository pattern which is maybe worth to have a look at. Not sure yet if I like it or not 😄.

1. It converts the Repository in a QueryBuilder itself:
https://github.com/mtarld/apip-ddd/blob/776527132d869f58380eb989b832db097158cbd1/src/Infrastructure/Shared/Doctrine/DoctrineRepository.php#L26-L28
2. Clone itself to avoid sideeffects:
https://github.com/mtarld/apip-ddd/blob/776527132d869f58380eb989b832db097158cbd1/src/Infrastructure/Shared/Doctrine/DoctrineRepository.php#L74-L79
3. Provides wither methods to filter itself:
https://github.com/mtarld/apip-ddd/blob/776527132d869f58380eb989b832db097158cbd1/src/Infrastructure/Library/Doctrine/DoctrineBookRepository.php#L41-L46
4. And is itself the list of objects:
https://github.com/mtarld/apip-ddd/blob/776527132d869f58380eb989b832db097158cbd1/src/Infrastructure/Shared/Doctrine/DoctrineRepository.php#L31-L40

The usage of such a repository looks like the following:

https://github.com/mtarld/apip-ddd/blob/776527132d869f58380eb989b832db097158cbd1/src/Application/Library/Query/FindBooksQueryHandler.php#L18-L28 

The problem I mostly see here is extendability, as the only way to extend it, would be overwriting the service in the project, so bundles could not add new filters which is sometimes required - could workaround by providing a withCustom filter which call subservices. Still an interesting pattern, the usage of it is a little bit uncommon. The "wither pattern" not sure if it is called this was is also a lot used in the psr-7 e.g. [Request->withMethod](https://github.com/Nyholm/psr7/blob/1461e07a0f2a975a52082ca3b769ca912b816226/src/RequestTrait.php#L66-L67) requires for immutability, which is required for a service.

Source: https://twitter.com/matarld/status/1512054080411217921

Which is inspired by: https://blog.mnavarro.dev/posts/repository-pattern-done-right/ -> https://blog.mnavarro.dev/the-repository-pattern-done-right

**Comment by alexander-schranz** [comment]
Had an interesting usecase in the comment bundle which currently @martinlagler had to implement first `3 comments` should be loaded and then onclick the next `100 comments` should be loaded. 

Currently we have in this RFC using used `limit` and `page` which did match our current usecase, but make things like that not possible. Instead the repository should go with `limit` and `offset` and if somebody uses `page/pageSize` they manually should convert that to `limit/offset` in the controller.

```diff
$page = 1;
$limit = 10;

-$repository->findBy(['page' => $page, 'limit' => $pageSize]);
+$repository->findBy(['limit' => ($page - 1) * $pageSize, 'offset' => $pageSize]);
```

This make the repository more flexible, the different controllers can decide then if paginated of limit/offset pagination is used for that endpoint.

**Comment by alexander-schranz** [comment]
Another topic I want to bring to this issue is about naming the filter parameters for `exluding`. So as example I have an `Organisation` and `Contact` and in an add overlay I want to get all `contacts` which are not in the organisation, currently I think we named that always `exclude-...`, in current repositories in the projects we did go with `excludeIds` or `excludeAccountId`, ... 

```php
$repository->findBy(['excludeAccountId' => $accountId]);
```

Other possibility would be "not" prefix, want to add this also to the discussion here how we handle the naming here:

```php
$repository->findBy(['notAccountId' => $accountId]);
```

**Update 2025-09-01**

As discussed with @Prokyonn we will go for negative filters with `exclude` prefix as in `excludeResourceId` in the new 3.0 RouteRepositoryInterface.

**Comment by RafaelKr** [comment]
Instead of using an `array` for the `$context` I would propose to type it with an Interface to be able to create Contexts from a class and specify the available options via constructor (or factory methods). If it somehow requires arbitrary array data the Context class can still contain an array property holding that (see the second last paragraph of my comment).

This helps us developers to quickly find available options without having to search the documentation. Also it prevents errors when these context properties are properly typed.
Another benefit is when the interface would change with later versions of Sulu it's much easier to find and adjust according code. It's also possible to deprecate properties or add them as needed.

TLDR: I wish for type safety where possible to:
- Prevent errors (for example typos in array keys or values)
- Have more self-documented code also helping us with autocompletion in an IDE supporting that
- Enabling static-analysis tools to be able to better check the code for possible errors.

### References

Shopware 6 uses a similiar concept of Context, but throughout a full request lifecycle. They implemented their own [Data Abstraction Layer](https://developer.shopware.com/docs/concepts/framework/data-abstraction-layer.html) where they do things in an elegant way (I don't propose to do this, this would open up a whole new world 😂). For every query you need to pass in a Context. See https://github.com/shopware/shopware/blob/88788d4055e7d4df01175ac4581f97c63f284cfe/src/Core/Framework/Context.php and https://github.com/shopware/shopware/blob/88788d4055e7d4df01175ac4581f97c63f284cfe/src/Core/Content/Media/SalesChannel/MediaRoute.php#L50-L59.

The context is generated per request and contains a lot of information like the current user, the currency, the current language chain (they can inherit from the base language if a translation is not provided), etc.

To see a real life example what they can do with the context is to look at how they version entities:
https://developer.shopware.com/docs/guides/plugins/plugins/framework/data-handling/versioning-entities.html#create-and-merge-version

Shopware has the concept of Sales Channels (they are basically the same as Sulu Webspaces). When a request is coming from the Frontend (e.g. Sales Channel) they use a special [SalesChannelContext](https://github.com/shopware/shopware/blob/88788d4055e7d4df01175ac4581f97c63f284cfe/src/Core/System/SalesChannel/SalesChannelContext.php#L25) inheriting from the base Context with some added properties and methods like the currently signed in customer, the customer group, etc.

Also they use this concept of contexts in other places. Those are not related to the default global Context I was just talking about. For example in their plugin system the plugins can have lifecycle methods which are called on specific events like install, uninstall, update, ...
https://developer.shopware.com/docs/guides/plugins/plugins/plugin-fundamentals/plugin-lifecycle.html#install
The method signature of the install lifecycle method looks like `public function install(InstallContext $installContext): void` and the InstallContext looks like this: https://github.com/shopware/shopware/blob/88788d4055e7d4df01175ac4581f97c63f284cfe/src/Core/Framework/Plugin/Context/InstallContext.php

Coming back to the initial paragraph where I spoke about arbitrary data: If you need an example how they add arbitrary data to a context you can look at [$context->addExtension](https://github.com/shopware/shopware/blob/2904165936edbb665ab35ecdc91680d20b50ebc6/src/Core/Framework/Struct/ExtendableTrait.php#L23). It's inherited via `Context extends Struct` and that uses the `ExtendableTrait`.

If you have any questions or would like to discuss this further let me know. I'll just joined the Sulu Slack community, DMs should be possible.

**Comment by alexander-schranz** [comment]
@RafaelKr I know what you mean but reminds me a lot o how to override things inside Sylius. You need atleast 3 services override Factory, Object, the class you want before have possibility for extending. I know type safety is great, but with a proper PHPStan configuration you can achieve the same typesafety. I know sure that some prefer such Value Objects but I think they are very beginner friendly, know Ondrej would say we need to teach them better ways but there is always a hard balance between beginner friendly and forcing specific patterns because we think we need to be strict.

So not a friend of it, want to tell that it totally make sense for projects which don't need flexibility / extendability like I think we need.
